rasa/shared/utils/schemas/events.py

```diff
@@ -0,0 +1,155 @@
+ENTITIES_SCHEMA = {
+    "type": "array",
+    "items": {
+        "type": "object",
+        "properties": {
+            "start": {"type": "integer"},
+            "end": {"type": "integer"},
+            "entity": {"type": "string"},
+            "confidence": {"type": "number"},
+            "extractor": {"type": ["string", "null"]},
+            "value": {},
+            "role": {"type": ["string", "null"]},
+            "group": {"type": ["string", "null"]},
+        },
+    },
+}
+
+INTENT = {
+    "type": "object",
+    "properties": {"name": {"type": "string"}, "confidence": {"type": "number"}},
+}
+
+EVENTS_SCHEMA = {
+    "type": "array",
+    "items": {
+        "type": "object",
+        "properties": {
+            "event": {"type": "string"},
+            "timestamp": {"type": ["number", "null"]},
+            "metadata": {"type": ["object", "null"]},
+        },
+        "required": ["event"],
+        "oneOf": [
+            {
+                "properties": {
+                    "event": {"const": "user"},
+                    "text": {"type": ["string", "null"]},
+                    "input_channel": {"type": ["string", "null"]},
+                    "message_id": {"type": ["string", "null"]},
+                    "parse_data": {
+                        "type": "object",
+                        "properties": {
+                            "text": {"type": "string"},
+                            "intent_ranking": {"type": "array", "items": INTENT},
+                            "intent": INTENT,
+                            "entities": ENTITIES_SCHEMA,
+                            "response_selector": {
+                                "type": "object",
+                                "properties": {
+                                    "all_retrieval_intents": {"type": "array"},
+                                    "default": {
```

Comment:
mhm, it might not always be `default` according to [this part](https://github.com/RasaHQ/rasa/blob/fe9f14919df4dcb454ba1052de344fae1466a6d1/rasa/nlu/selectors/response_selector.py#L451). Is there a simple solution in jsonschema for dynamic keys? Otherwise I'd skip that for now

Reply:
I was only able to find resource on defining [regular expressions](https://json-schema.org/understanding-json-schema/reference/object.html#pattern-properties) for keys - would this be suitable?

Reply:
yeah, that would certainly work 👍🏻

Reply:
> So it seems it's a dead end and I should not implement patternProperties in the schema? WDYT?

Sounds good for me for now. We can still revisit that in the future 👍🏻

Reply:
I've found a way to make this work actually 💡 I used `oneOf` and the tests pass - the caveat is that if `all_retrieval_intents` is a dict instead of list, then the schema won't fail because the `patternProperties` subschema will be valid for this key name 😞